README.md
# ubuntu-18.04-rke

Vagrant Box with Ubuntu 18.04 and rke

## Base image

Used base image [elegoev/ubuntu-18.04-docker](https://app.vagrantup.com/elegoev/boxes/ubuntu-18.04-docker)

## Directory Description

| directory | description                                          |
|-----------|------------------------------------------------------|
| inspec    | inspec test profiles with controls                   |
| packer    | packer build, provisioner and post-processor scripts |
| test      | test environment for provision & inspec development  |

## Configuration

### Vagrant Cloud

- [elegoev/ubuntu-18.04-rke](https://app.vagrantup.com/elegoev/boxes/ubuntu-18.04-rke)

### Useful Vagrant Plugins

- [vagrant-disksize](https://github.com/sprotheroe/vagrant-disksize)
- [vagrant-hosts](https://github.com/oscar-stack/vagrant-hosts)
- [vagrant-secret](https://github.com/tcnksm/vagrant-secret)
- [vagrant-vbguest](https://github.com/dotless-de/vagrant-vbguest)
- [vagrant-serverspec](https://github.com/vvchik/vagrant-serverspec)
- [vagrant-vmware-esxi](https://github.com/josenk/vagrant-vmware-esxi)

### Using rke

#### Vagrantfile

    Vagrant.configure("2") do |config|

        $basebox_name="ubuntu-18.04-rke-test"
        $basebox_hostname="ubuntu-1804-rke-test"
        $src_image_name="elegoev/ubuntu-18.04-rke"
        $vb_group_name="basebox-rke-test"

        config.vm.define "#{$basebox_name}" do |machine|
          machine.vm.box = "#{$src_image_name}"
    
          # define guest hostname
          machine.vm.hostname = "#{$basebox_hostname}"

          machine.vm.provider "virtualbox" do |vb|
            vb.name = $basebox_name
            vb.cpus = 1
            vb.customize ["modifyvm", :id, "--memory", "1024" ]
            vb.customize ["modifyvm", :id, "--natdnshostresolver1", "on"]
            vb.customize ["modifyvm", :id, "--natdnsproxy1", "on"]
            vb.customize ["modifyvm", :id, "--groups", "/#{$vb_group_name}" ]
            vb.customize ["modifyvm", :id, "--vram", 256 ]
          end

        end   

    end

#### Help

    rke --help

### Referenzen

- [Rancher Kubernetes Engine (RKE)](https://rancher.com/products/rke/)
- [Github RKE](https://github.com/rancher/rke)
- [Creating a vSphere Cluster](https://rancher.com/docs/rancher/v2.x/en/cluster-provisioning/rke-clusters/node-pools/vsphere/)
- [K8s Ingress Controllers](https://rancher.com/docs/rke/latest/en/config-options/add-ons/ingress-controllers/)
- [Metallb kubernetes-rke-baremetal](https://github.com/mimizone/kubernetes-rke-baremetal/tree/master/metallb)

### Versioning

Repository follows sematic versioning  [![semantic versioning](https://img.shields.io/badge/semver-2.0.0-green.svg)](http://semver.org)

### Changelog

For all notable changes see [CHANGELOG](https://github.com/elegoev/basebox-ubuntu-18.04-rke/blob/master/CHANGELOG.md)

### License

Licensed under The MIT License (MIT) - for the full copyright and license information, please view the [LICENSE](https://github.com/elegoev/basebox-ubuntu-18.04-rke/blob/master/LICENSE) file.

### Issue Reporting

Any and all feedback is welcome.  Please let me know of any issues you may find in the bug tracker on github. You can find it [here.](https://github.com/elegoev/basebox-ubuntu-18.04-rke/issues)


CHANGELOG.md
# Change Log

## 1.0.x

**Implemented enhancements:**

- rke installation

## 2.0.x

**Implemented enhancements:**

- Packer build implemented


packer/provisioners/create-ssh-key.sh
#!/bin/bash

# set color
GREEN='\033[0;32m'
RED='\033[0;31m'
BLUE='\033[0;34m'
NC='\033[0m'

server_hostname=$(hostname)

local_ssh_path="/home/vagrant/.ssh"
root_ssh_path="/root/.ssh"
ssh_privkey_filename="id_rsa"
ssh_pubkey_filename="id_rsa.pub"
ssh_authorized_keys="authorized_keys"
ssh_known_hosts="known_hosts"

# create ssh key
sudo ssh-keygen -t rsa -C "developer@test.org" -b 4096 -N "" -f $local_ssh_path/$ssh_privkey_filename
sudo chmod 600 $local_ssh_path/$ssh_privkey_filename
sudo chmod 644 $local_ssh_path/$ssh_pubkey_filename

# add public key to authorized keys
sudo cat $local_ssh_path/$ssh_pubkey_filename >> $local_ssh_path/$ssh_authorized_keys

# add known host
sudo ssh-keyscan -H localhost >> $local_ssh_path/$ssh_known_hosts

# set ownership
sudo chown -R vagrant:vagrant $local_ssh_path

# copy ssh key for root
if [ -d "$root_ssh_path" ]
then
  sudo rm -fr $root_ssh_path
fi
sudo cp -r $local_ssh_path /root


packer/provisioners/base-provisioning.sh
#!/bin/bash

application_file_path="/vagrant/installed-application.md"
rancher_cluster_dir="/home/vagrant/rancher-rke-cluster"

# set variables
rke_version="1.2.5"
rke_filename="rke_linux-amd64"
rke_bin_dir="/usr/local/bin"

# get rke image
sudo wget -q -O /tmp/$rke_filename https://github.com/rancher/rke/releases/download/v$rke_version/$rke_filename
sudo mv /tmp/$rke_filename $rke_bin_dir/rke
sudo chmod +x $rke_bin_dir/rke

# install kubectl
sudo snap install kubectl --classic

# install helm
sudo snap install helm --classic

# create rancher cluster directory
sudo mkdir ${rancher_cluster_dir}
sudo chown vagrant:vagrant ${rancher_cluster_dir}

# copy cluster config
ls -al /home/vagrant/
ls -al /home/vagrant/files-prov
sudo cp /home/vagrant/files-prov/rke/rancher-cluster.yml ${rancher_cluster_dir}/rancher-cluster.yml
sudo chown vagrant:vagrant ${rancher_cluster_dir}/rancher-cluster.yml

# create cluster
cd ${rancher_cluster_dir}
rke up --config ./rancher-cluster.yml
sudo chown vagrant:vagrant *

# copy kube config to $HOME/.kube/config
kube_config_dir="/home/vagrant/.kube"
if [ ! -d "$kube_config_dir" ]; then
  sudo mkdir -p $kube_config_dir
fi
sudo cp kube_config_rancher-cluster.yml $kube_config_dir/config
sudo chown vagrant:vagrant $kube_config_dir/config
sudo chmod 0600 $kube_config_dir/config

# copy .kube for root
sudo cp -r $kube_config_dir /root

# set version
DOCKER_VERSION=$(sudo docker version --format '{{.Server.Version}}')
KUBECTL_VERSION=$(sudo snap info kubectl | grep installed | awk  '{print $2}')
HELM_VERSION=$(sudo snap info helm | grep installed | awk  '{print $2}')
RKE_VERSION=$(rke --version | awk  '{print $3}' | tr --delete v)
echo "# Installed application "  > $application_file_path
echo "***                     " >> $application_file_path
echo "> Docker: $DOCKER_VERSION" >> $application_file_path
echo "> Kubectl: $KUBECTL_VERSION" >> $application_file_path
echo "> Helm: $HELM_VERSION" >> $application_file_path
echo "> rke: $RKE_VERSION" >> $application_file_path




packer/post-processors/create-git-tag.ps1
# create git tag

$src_image_version = $env:IMAGE_VERSION

$vagrant_file_template = '
Vagrant.configure("2") do |config|

    $basebox_name="ubuntu-18.04-rke-test"
    $basebox_hostname="ubuntu-1804-rke-test"
    $src_image_name="elegoev/ubuntu-18.04-rke"
    $vb_group_name="basebox-rke-test"

    config.vm.define "#{$basebox_name}" do |machine|
      machine.vm.box = "#{$src_image_name}"
      machine.vm.box_version = "$image_version"
    
      # define guest hostname
      machine.vm.hostname = "#{$basebox_hostname}"

      machine.vm.provider "virtualbox" do |vb|
        vb.name = $basebox_name
        vb.cpus = 1
        vb.customize ["modifyvm", :id, "--memory", "1024" ]
        vb.customize ["modifyvm", :id, "--natdnshostresolver1", "on"]
        vb.customize ["modifyvm", :id, "--natdnsproxy1", "on"]
        vb.customize ["modifyvm", :id, "--groups", "/#{$vb_group_name}" ]
        vb.customize ["modifyvm", :id, "--vram", 256 ]
      end

    end   

end
'

# create Vagrantfile for download test
$vagrant_file_expanded = $vagrant_file_template.Replace('$image_version', "$src_image_version")
Set-Content -Path './test/download/Vagrantfile' -Value "$vagrant_file_expanded"

# create git tag
git add --all
git commit -m "new basebox version $src_image_version"
git push 
git tag $src_image_version
git push --tags





